ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,468 @@
+    // make a pre check for conflict between the accounts to update and the accounts in cache. Will fail this update
+    // the update operation if any conflict exists. There is a slight chance that the account to update conflicts with
+    // the accounts in the local cache, but does not conflict with those in the helixPropertyStore. This will happen
+    // if some accounts are updated but the local cache is not refreshed.
+    Pair<Account, Account> conflictAccountPair = getConflictAccountWithCache(accounts);
+    if (conflictAccountPair != null) {
+      // @todo should we just log the case without throwing an exception? This case is indeed possible and may not be
+      // @todo caused by a programming error.
+      throw new IllegalArgumentException(
+          "Account to update conflicts with the existing account. accountId=" + conflictAccountPair.getFirst().getId()
+              + " accountName=" + conflictAccountPair.getFirst().getName() + " conflicts with existing accountId="
+              + conflictAccountPair.getSecond().getId() + " accountName=" + conflictAccountPair.getSecond().getName());
+    }
+
+    boolean res = helixStore.update(FULL_ACCOUNT_METADATA_PATH, currentData -> {
+      ZNRecord newRecord;
+      if (currentData == null) {
+        logger.debug(
+            "ZNRecord does not exist on path={} in HelixPropertyStore when updating accounts. Creating a new ZNRecord.",
+            FULL_ACCOUNT_METADATA_PATH);
+        newRecord = new ZNRecord(String.valueOf(ZN_RECORD_ID));
+      } else {
+        newRecord = new ZNRecord(currentData);
+      }
+      Map<String, String> accountMap = newRecord.getMapField(ACCOUNT_METADATA_MAP_KEY);
+      if (accountMap == null) {
+        logger.debug("AccountMap does not exist in ZNRecord when updating accounts. Creating a new accountMap");
+        accountMap = new HashMap<>();
+      }
+      AccountMapPair accountMapPair = new AccountMapPair(accountMap);
+      Pair<Account, Account> conflictPair = getConflictAccount(accounts, accountMapPair);
+      // if there is any conflict with the existing record, or any exception occurs during building the new ZNRecord,
+      // the current policy is not to make any update.
+      if (conflictPair != null) {
+        logger.error("AccountId={} accountName={} conflicts with remote accountId={} accountName={}. Stopping update "
+                + "operation and no update will be made for accounts={}", conflictPair.getFirst().getId(),
+            conflictPair.getFirst().getName(), conflictPair.getSecond().getId(), conflictPair.getSecond().getName(),
+            accounts);
+      } else {
+        boolean noException = true;
+        for (Account account : accounts) {
+          try {
+            accountMap.put(String.valueOf(account.getId()), account.toJson().toString());
+          } catch (Exception e) {
+            logger.error("Unexpected exception when updating accountId={} accountName={} to ZNRecord={}",
```

Comment:
Throw a `RuntimeException` which will be captured by Helix and eventually make the return value to be `false`.